File: src/main/java/com/evilvaca/loteca/scrapers/TransferMarktLeagueScraper.java
package com.evilvaca.loteca.scrapers;

import com.evilvaca.loteca.domain.Match;
import com.evilvaca.loteca.domain.Round;
import lombok.RequiredArgsConstructor;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class TransferMarktLeagueScraper {

    public static final String HOST_URL = "https://www.transfermarkt.com.br";

    private static final String TOURNAMENT_URL = HOST_URL + "/{tournamentCodeDescription}/spieltag/wettbewerb/{tournamentCode}/saison_id/{seasonCode}/spieltag/{roundCode}";

    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("ddMMyyHHmm");

    private final RestTemplate restTemplate;

    public List<Match> getMatchList(Round round) {
        return getDocument(TOURNAMENT_URL,
                round.getSeason().getTournament().getCodeDescription(),
                round.getSeason().getTournament().getCode(),
                round.getSeason().getCode(),
                round.getCode())
                .select("table tbody tr.table-grosse-schrift").stream()
                .map(match -> getMatchDetails(match.parents().get(2).select("div.footer a").attr("href")))
                .collect(Collectors.toList());
    }

    private Match getMatchDetails(String url) {
        String matchId = url.substring(url.lastIndexOf('/') + 1);
        Document doc = getDocument(HOST_URL + url);

        String homeTeam = doc.select(" div.sb-team a.sb-vereinslink").get(0).attr("href")
                .replaceFirst("[^0-9]*", "").split("/")[0];

        String awayTeam = doc.select(" div.sb-team a.sb-vereinslink").get(1).attr("href")
                .replaceFirst("[^0-9]*", "").split("/")[0];

        boolean hasScore = doc.select(".sb-endstand").text().matches(".*[0-9]:[0-9].*");

        String homeScore = hasScore ? doc.select(".sb-endstand").text().split("[: ]")[0] : null;
        String awayScore = hasScore ? doc.select(".sb-endstand").text().split("[: ]")[1] : null;
        LocalDateTime matchDateTime = LocalDateTime.parse(doc.select(".sb-datum").text()
                .replaceFirst(".+?\\|", "")
                .replaceAll("[^0-9]", ""), DATE_TIME_FORMATTER);
        String stadium = doc.select(".sb-zusatzinfos span a").text();

        return Match.builder()
                .id(Long.parseLong(matchId))
                .homeId(Long.parseLong(homeTeam))
                .awayId(Long.parseLong(awayTeam))
                .homeScore(hasScore ? Integer.parseInt(homeScore) : null)
                .awayScore(hasScore ? Integer.parseInt(awayScore) : null)
                .startAt(matchDateTime)
                .stadium(stadium)
                .build();
    }

    private Document getDocument(String url, Object... uriVariables) {
        final HttpHeaders headers = new HttpHeaders();
        headers.set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.54 Safari/537.36");
        ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, new HttpEntity<String>(headers), String.class, uriVariables);
        return Jsoup.parse(response.toString());
    }
}

File: scraper/transfermarkt_league_scraper.go
package scraper

import (
	"fmt"
	"net/http"
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/giancarlobastos/loteca-backend/domain"
	"github.com/nleeper/goment"
	"golang.org/x/net/html"
)

const ()

type TransferMarktLeagueScraper struct {
}

func NewTransferMarktLeagueScraper() *TransferMarktLeagueScraper {
	return &TransferMarktLeagueScraper{}
}

func (c *TransferMarktLeagueScraper) GetMatchList(round domain.Round) (*[]domain.Match, error) {
	url := fmt.Sprintf("https://www.transfermarkt.com.br/%s/spieltag/wettbewerb/%s/saison_id/%s/spieltag/%s",
		round.Season.Competition.CodeName,
		round.Season.Competition.Code,
		round.Season.Code,
		round.Code)

	doc := c.getDocument(url)
	matches := make([]domain.Match, 10)

	doc.Find("div.footer a[href*='bericht/index/spielbericht/']").Each(
		func(i int, s *goquery.Selection) {
			matchDetailsUrl, _ := s.Attr("href")
			match, _ := c.getMatchDetails(matchDetailsUrl)
			matches = append(matches, *match)
		})

	return &matches, nil
}

func (c *TransferMarktLeagueScraper) getMatchDetails(url string) (*domain.Match, error) {
	doc := c.getDocument("https://www.transfermarkt.com.br" + url)
	match := domain.Match{
		HomeScore: -1,
		AwayScore: -1,
	}

	urlSplit := strings.Split(url, "/")
	matchId, _ := strconv.ParseUint(urlSplit[len(urlSplit)-1], 10, 32)
	match.Id = uint32(matchId)

	teamUrls := doc.Find("div.sb-team  a.sb-vereinslink").Nodes

	homeTeamUrl := c.attr(teamUrls[0], "href")
	homeTeamUrlSplit := strings.Split(homeTeamUrl, "/")
	homeTeamId, _ := strconv.ParseUint(homeTeamUrlSplit[4], 10, 32)
	match.HomeId = uint32(homeTeamId)

	awayTeamUrl := c.attr(teamUrls[1], "href")
	awayTeamUrlSplit := strings.Split(awayTeamUrl, "/")
	awayTeamId, _ := strconv.ParseUint(awayTeamUrlSplit[4], 10, 32)
	match.AwayId = uint32(awayTeamId)

	scorePattern := regexp.MustCompile("[^0-9]*([0-9]+):([0-9]+).*")
	score := doc.Find(".sb-endstand").Text()
	hasScore := scorePattern.MatchString(score)

	if hasScore {
		match.HomeScore, _ = strconv.Atoi(scorePattern.FindStringSubmatch(score)[1])
		match.AwayScore, _ = strconv.Atoi(scorePattern.FindStringSubmatch(score)[2])
	}

	datePattern := regexp.MustCompile(`[^0-9]*([0-9]{2}/[0-9]{2}/[0-9]{2})[^0-9]*([0-9]{2}:[0-9]{2})`)
	dateSplit := datePattern.FindStringSubmatch(doc.Find(".sb-datum").Text())
	formatter, _ := goment.New(dateSplit[1]+dateSplit[2], "DD/MM/YYHH:mm")

	match.StartAt = formatter.ToTime()
	match.Stadium = doc.Find(".sb-zusatzinfos span a").Text()

	return &match, nil
}

func (c *TransferMarktLeagueScraper) getDocument(url string) *goquery.Document {
	fmt.Println(url)
	client := &http.Client{}
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.54 Safari/537.36")

	res, _ := client.Do(req)
	defer res.Body.Close()

	doc, _ := goquery.NewDocumentFromReader(res.Body)
	return doc
}

func (c *TransferMarktLeagueScraper) attr(n *html.Node, key string) string {
	if n.Type == html.ElementNode && n.Data == "a" {
		for _, a := range n.Attr {
			if a.Key == key {
				return a.Val
			}
		}
	}
	return ""
}
